### PR TITLE
Fix #3566 use an hardware timer for software PWM stability

### LINKED ISSUE #3566: xd60 backlight issues
Hi, 

I'm new to QMK, but I was able to produce my own keymap and flash it to the keyboard.

Unfortunately, I have some backlight issues:

* BL_TOGG doesn't work, but BL_STEP or BL_INC/DEC works correctly
* when in lower backlight levels (like 1 to 4), when typing the backlight frequently flashes (going brighter), which is really annoying. It gets worst if more than one key is pressed at the same time (like pressing QWERTYU, really makes it flashing 100% of the time).

I double-checked all the soldering, and everything seems good.

Unfortunately I haven't used the original firmware (TKG based) to see if it exhibits the same thing.

While looking at the QMK code, I see that backlight uses B6 so hardware PWM. 

It doesn't seem to be the case for the original firmware (see https://github.com/kairyu/tmk_keyboard_custom/blob/581da2a49c46739d0dfed636f7f201b7d60b77e1/keyboard/gh60/backlight.c) which for GH60_REV_ABC (the recommended vendor settings when flashing the original firmware) enables the software PWM.

Could that be the issue?
Any idea on how to fix the flickering?

Thanks for any help,

### PR DESCRIPTION
**Pull request by masterzen**
With my XD60, I noticed that when typing the backlight was flickering (see issue #3566).

The XD60 doesn't have the backlight wired to a hardware PWM pin (too bad), unlike the dz60.

I assumed it was a timing issue in the matrix scan that made the PWM lit the LED a bit too longer. I verified it because the more keys that were pressed, the more lighting flicker I observed.

This patch removes the matrix scan based PWM computation, and instead adds a timer interrupt that ticks the backlight PWMs, leading to a better time stability, and removing all flickering on my keyboard :)

I think this patch might not work with some other software PWM based keyboards as I'm using Timer1B (and have only this xd60 to test with).

I'd be more than happy to amend this patch with a more generic fix or a better solution if needed :)

Thanks,
Brice

### CODE REVIEW

**Comment by drashna** [comment]
Timer 1 is used by audio, in some cases, so this may cause issues.

**Comment by masterzen** [comment]
@drashna ,

Thanks for the input!
I've modified my PR to disable the hardware timer whenever audio is enabled (do you think that's enough?).

I've also fixed the mt40 compilation. Hopefully the travis tests will pass now.

Thanks,
--
Brice

**Comment by drashna** [comment]
I think that a warning should be generated, actually.  

That way, there isn't any confusion about why it's having issues if you have audio enabled. 

Though, to clarify, it's the B pins (B5, B6, B7) that use Timer 1.  So you could detect if `B5_AUDIO`, `B6_AUDIO` or `B7_AUDIO` is defined, and then error out then, rather than if audio is enabled in general

**Comment by masterzen** [comment]
Hi @drashna 

I did something a bit differently than what you suggested. If any of `B5_AUDIO`, `B6_AUDIO`, `B7_AUDIO` is enabled, we'll use Timer 3. If it's any of `C6_AUDIO`, `C5_AUDIO`, or `C4_AUDIO`,  (or nothing) then we'll use Timer 1. If both timers are in use with audio then we won't use this feature and revert to matrix scan PWM computation.

I think this is the best solution, WDYT?

I've tested both timers on my xd60, but since I have no keyboard with audio it's possible I've missed something.

Thanks!
Brice

**Comment by drashna** [comment]
That sounds good! 

@jackhumbert ?

**Comment by masterzen** [comment]
@drashna , I've added a small documentation change, but since I'm not an English native speaker, you might want to proof-read it (or propose some rewriting).

File: docs/feature_backlight.md [hunk 1]
```diff
@@ -30,7 +30,7 @@ You should then be able to use the keycodes below to change the backlight level.
 
 This feature is distinct from both the [RGB underglow](feature_rgblight.md) and [RGB matrix](feature_rgb_matrix.md) features as it usually allows for only a single colour per switch, though you can obviously use multiple different coloured LEDs on a keyboard.
 
-Hardware PWM is only supported on certain pins of the MCU, so if the backlighting is not connected to one of them, a software implementation will be used, and backlight breathing will not be available. Currently the supported pins are `B5`, `B6`, `B7`, and `C6`.
+Hardware PWM is only supported on certain pins of the MCU, so if the backlighting is not connected to one of them, a software PWM implementation based on a hardware timer will be used, and backlight breathing will not be available. Currently the supported pins for hardware PWM are `B5`, `B6`, `B7`, and `C6`. If audio is enabled while backlight uses software PWM, and depending on the audio pins used (Cx or Bx), the hardware timer might be either Timer 1 or Timer 3. If the audio uses both timers, then the backlight will resort to a matrix scan scheduling which might give backlight lighting artifacts (flickering).
```
Comment (drashna) [inline docs/feature_backlight.md]: C4 and C5 are probably supported too, as they support Timer 3.

Also, it may be a good idea to clarify that the timers are only supported on certain pins, too. (like how audio lists them: https://docs.qmk.fm/#/feature_audio)

Comment (masterzen) [reply docs/feature_backlight.md]: @drashna , I've reworded completely this part of the documentation and inserted tables to show what timers are in use in which configuration of backlight and audio. I think this makes it more clear what is possible and what is not.

I double checked the code, and QMK only supports hardware PWM based backlight on B5, B6, B7 and C6. I did a quick checkin the source code and no QMK supported keyboards are using other pins on the C port besides C6. There are other backlight pin (like my xd60 which uses F5), but those don't have a hardware PWM.

I couldn't find a way to produce an html version of the documentation, so I wasn't able to see if everything renders correctly (does it use jekyll ?).

Thanks,
Brice

**Comment by masterzen** [comment]
@drashna were you able to proof-read the latest version of the documentation I pushed yesterday 
?

Let me know if there's any modifications I should add.

Thanks,
Brice

**Comment by drashna** [comment]
Documentation looks good to me

**Comment by masterzen** [comment]
While thinking about this, I have a better idea that can reuse most of the hardware PWM code and will allow breathing to work in software PWM mode. I'll work on this during the week-end and push there when ready.

**Comment by masterzen** [comment]
I've pushed the latest changes which works well for my xd60 implementing this last idea:

* reuses most of hardware PWM code
* compatible with Audio if only one pin is in use by Audio
* supports breathing in software mode
* doesn't flicker when typing like software mode
* contains documentation and several code comments to explain it
* support for multiple backlight pins  (allows for instance to control the brightness of other leds to match the backlight leds, useful when Caps Lock is mapped as another key and you want the caps lock led to be part of the backlight). This feature is only available in pure software mode and this new hybrid mode, but not hardware mode.

I haven't unfortunately been able to test if hardware PWM mode is still working (I don't see why it wouldn't), because I don't have any keyboards with this mode. I'd appreciate if someone could test that for me.

Thanks for taking a look, I'm ready for comments and suggestions :)

**Comment by mikeinkl** [comment]
Just came here to say thanks. The current implementation of software PWM is quite buggy (Toggle key not working, flashing during suspend, doesn't work for more than 5 backlight levels). I managed to fix all that but couldn't fix the flickering during typing, so was about to cross traces on my KC60se, when I saw this.

**Comment by masterzen** [comment]
Hey @jackhumbert, @skullydazed or @fredizzimo , @drashna added you as potential reviewers for this PR.

I really think this PR would be a great addition to QMK and would solve the software PWM backlight flickering issue that happens on software PWM boards.

I'd love if any of you would be able to review or test this PR and let me know if anything should be amended, before this PR gets buried under tons of other PR :)
I'm really sorry for begging :)

Many thanks!

**Comment by drashna** [comment]
It would definitely be nice to test this.  Unfortunately, I don't have any keyboards with backlighting (or the ones that do, don't have the hardware installed). So I can't test this personally. 
And the hardware related stuff (like timers) is above my head (so far).  

And the others are busy, so haven't had a chance to get to this PR.  However, we are trying to change our workflow to get to PRs like this much faster.

**Comment by masterzen** [comment]
@drashna , it wasn't meant as criticism at all :) and I'm sorry if it sounded like that (English is not my native language).

I know it is very hard to manage an open source project of that scale, that it takes a lot of time and efforts for you all. I'm grateful such project exists to power my keyboards!

I see a handful of timer based backlight code copy/pasted between keyboard definitions in QMK (search for `timer1PWMSetup` too see what I mean). This PR would allow all those keyboards to benefit from a core implementation :)

Anyway, thanks for all!

**Comment by drashna** [comment]
No, no, I didn't take it that way, at all.  I just wanted to let you know that's why.  

Basically, too much stuff, not enough time. :)
And like I said, we're actively trying to get more efficient about this sort of thing. So.... :)

**Comment by a-chol** [comment]
Hey, I read your comment in my PR (4171) : should I be able to test your changes to confirm it works? Would that move your PR forward?

**Comment by masterzen** [comment]
@a-chol that would be awesome if you could try it and report here how it worked for your board!

To use the software PWM assisted by a timer (akin to the code in your PR) you just need to define the `BACKLIGHT_PIN` as usual along with `BACKLIGHT_BREATHING` if you want breathing.

Many thanks!

**Comment by a-chol** [comment]
Si I tried building your branch with my keyboard definition, to which I remove the custom backlight driver, set BACKLIGHT_CUSTOM_DRIVER = no. I managed to build, but the backlight doesn't turn up. 
My keyboard is based on an ATMega32A. Looking at the datasheet, since my backlight is connected to pin D4 and pin D4 supports the OC1B, I should be able to use full hardware PWM.
I think your code is specific to ATMega32U4 and the pin layout being different, the code needs to be modified to work with different MCU.

**Comment by masterzen** [comment]
@a-chol , `D4` is not supported by the QMK hardware PWM implementation which my patch doesn't touch. But it should work on software PWM, irrespective of the timer used.

If you see the "Using hardware timer 1 with software PWM", then it is using my code. If that doesn't work for your board, try to disable breathing first. Then check the number of BACKLIGHT_STEP defined (I have 6 defined on my board), maybe there's not enough.

I checked the Atmega32A datasheet and the timer1 seems to work exactly in the same way as the Atmega32U4 16bits timers. I double-checked the initialization of the 16bits timer and it seems to be correct for both MCU.

If you want to try full hardware PWM on pin D4 (this is almost the same code as my software PWM, so this could give hints if it doesn't work), you just need to add a new pin definition for D4 [in backlight.c](https://github.com/qmk/qmk_firmware/pull/3615/files#diff-7cc280270d4afece33b8b15ed53abebbL974) that would look like this (if D4 is on OCR1B):

~~~c
...
#elif BACKLIGHT_PIN == D4
#  define HARDWARE_PWM
#  define TCCRxA TCCR1A
#  define TCCRxB TCCR1B
#  define COMxx1 COM1B1
#  define OCRxx  OCR1B
#  define ICRx   ICR1
...
~~~ 

If that works for you, then there's no reason the software PWM doesn't work since it uses the exact same timer definition (Fast PWM mode 14), except it doesn't control the output pins (which is done in an interrupt).
Let me know how it goes, we'll see how we can find the problem.

Thanks for trying!

**Comment by a-chol** [comment]
I tried modifying the section you suggested in roughly the same way and got hardware PWM working. I don't get why the software PWM doesn't work either. I'll try to spend a little more time on this in the coming days.

**Comment by a-chol** [comment]
I've replace your bit-fiddling technics to deduce the PORTx and DDRx from the backlight_pin in quantum.c with hardcoded PORTD and DDRD, and I got the backlight working.
Looking at the includes for the AVR toolchain, PORTD is defined as 0x0B for the ATMega32U4, which works well with your method, but for the ATMega32A, it is defined as 0x12.

**Comment by masterzen** [comment]
@a-chol you're completely right!

The combination of [Quantum common_config.h](https://github.com/qmk/qmk_firmware/blob/master/quantum/config_common.h#L28) and the trick of `_SFR_IO8((backlight_pin >> 4) + 1) |= _BV(backlight_pin & 0xF);` (used in place for instance of `DDRF |= (1 << PINF5)`) works only for Atmega32U4 for the moment.

The reason is that the pin value in [Quantum common_config.h](https://github.com/qmk/qmk_firmware/blob/master/quantum/config_common.h#L28) are setup only for the Atmega32U4.

So I wondered how it could simply work for matrix scan reading in keyboards that don't use an Atmega32U4 like yours of the `jj40`. The answer is that keyboards redefined their own matrix scan using `DDRx` and `PORTx` macro.

@jackhumbert has created PR #4015 to solve this problem, in which he provides a multi MCU `common_config.h` along with the correct macros to set/unset pins level instead of hardcoding pins per keyboards.

I'm going to see if I can rebase on top of #4015 and use those macros, which makes more sense than reinventing the wheel just for the backlight. Then if you have time that'd be great if you could retest with your board.

**Comment by masterzen** [comment]
@a-chol you're completely right!

The combination of [Quantum common_config.h](https://github.com/qmk/qmk_firmware/blob/master/quantum/config_common.h#L28) and the trick of `_SFR_IO8((backlight_pin >> 4) + 1) |= _BV(backlight_pin & 0xF);` (used in place for instance of `DDRF |= (1 << PINF5)`) works only for Atmega32U4 for the moment.

The reason is that the pin value in [Quantum common_config.h](https://github.com/qmk/qmk_firmware/blob/master/quantum/config_common.h#L28) are setup only for the Atmega32U4.

So I wondered how it could simply work for matrix scan reading in keyboards that don't use an Atmega32U4 like yours of the `jj40`. The answer is that keyboards redefined their own matrix scan using `DDRx` and `PORTx` macro.

@jackhumbert has created PR #4015 to solve this problem, in which he provides a multi MCU `common_config.h` along with the correct macros to set/unset pins level instead of hardcoding pins per keyboards.

I'm going to see if I can rebase on top of #4015 and use those macros, which makes more sense than reinventing the wheel just for the backlight. Then if you have time that'd be great if you could retest with your board.

**Comment by masterzen** [comment]
@a-chol you're completely right!

The combination of [Quantum common_config.h](https://github.com/qmk/qmk_firmware/blob/master/quantum/config_common.h#L28) and the trick of `_SFR_IO8((backlight_pin >> 4) + 1) |= _BV(backlight_pin & 0xF);` (used in place for instance of `DDRF |= (1 << PINF5)`) works only for Atmega32U4 for the moment.

The reason is that the pin value in [Quantum common_config.h](https://github.com/qmk/qmk_firmware/blob/master/quantum/config_common.h#L28) are setup only for the Atmega32U4.

So I wondered how it could simply work for matrix scan reading in keyboards that don't use an Atmega32U4 like yours of the `jj40`. The answer is that keyboards redefined their own matrix scan using `DDRx` and `PORTx` macro.

@jackhumbert has created PR #4015 to solve this problem, in which he provides a multi MCU `common_config.h` along with the correct macros to set/unset pins level instead of hardcoding pins per keyboards.

I'm going to see if I can rebase on top of #4015 and use those macros, which makes more sense than reinventing the wheel just for the backlight. Then if you have time that'd be great if you could retest with your board.

**Comment by masterzen** [comment]
@a-chol you're completely right!

The combination of [Quantum common_config.h](https://github.com/qmk/qmk_firmware/blob/master/quantum/config_common.h#L28) and the trick of `_SFR_IO8((backlight_pin >> 4) + 1) |= _BV(backlight_pin & 0xF);` (used in place for instance of `DDRF |= (1 << PINF5)`) works only for Atmega32U4 for the moment.

The reason is that the pin value in [Quantum common_config.h](https://github.com/qmk/qmk_firmware/blob/master/quantum/config_common.h#L28) are setup only for the Atmega32U4.

So I wondered how it could simply work for matrix scan reading in keyboards that don't use an Atmega32U4 like yours of the `jj40`. The answer is that keyboards redefined their own matrix scan using `DDRx` and `PORTx` macro.

@jackhumbert has created PR #4015 to solve this problem, in which he provides a multi MCU `common_config.h` along with the correct macros to set/unset pins level instead of hardcoding pins per keyboards.

I'm going to see if I can rebase on top of #4015 and use those macros, which makes more sense than reinventing the wheel just for the backlight. Then if you have time that'd be great if you could retest with your board.

**Comment by a-chol** [comment]
I tried merging your PR on top of #4015 and it works fine.
One weird thing though is that the intensity seems to be reversed : it decreases with BL_INC and inversely. I think it's linked to BACKLIGHT_ON_STATE, but I'm not sure why this inverts the intensity by default. Is it used to turn on backlight at startup?
Also, I tested with 2 backlight pins, and it works fine.

**Comment by a-chol** [comment]
If I'm not mistaken, your backlight_on function turns OFF lights when BACKLIGHT_ON_STATE==0, inversely to what you described. By the way, my keyboard config doesn't touch BACKLIGHT_ON_STATE at all, so the default (=0) is applied. Isn't the logic reversed in your non-hardware backlight_on/backlight_off functions?

**Comment by masterzen** [comment]
@a-chol, yes I re-read the code yesterday, and it seems I've got my explanation wrong, sorry :(

Note that my code reused what was used before for the software PWM, so I guess it should work :)

I checked and for instance the following regular AVR code:
~~~c
PORTF &= ~(1 << PF5); // PF5 goes low
PORTF |= (1 << PF5); // PF5 goes high 
~~~
which translates to in terms of `_SFR_IO8` macro:
~~~c
_SFR_IO8((F5 >> 4) + 2) |= _BV(F5 & 0xF); // F5 high
_SFR_IO8((F5 >> 4) + 2) &= ~_BV(F5 & 0xF); // F5 low
~~~
which BTW is now in the QMK code base as `writePinHigh` and `writePinLow` (I'm going to update my PR to use that instead).

So that means when `BACKLIGHT_ON_STATE == 0` we put the pin __low__ to turn on the leds, and without that we put the pin __high__ to turn on the leds. Unfortunately `BACKLIGHT_ON_STATE` is not documented. The only reference I could find is PR #1073.

I'm not that good in electronics, but from what I understand the MCU pin is not able to power the leds directly, instead it is connected to a MOSFET transistor (itself connected to Vcc, GND and the leds). Depending on how this last one is wired, sometimes the transistor powers the led when the command pin is low, sometimes when it is high, so you have to define `BACKLIGHT_ON_STATE` accordingly (1 for powered on high, 0 for powered on low).

Can you check on your board what level is required to power the leds (it probably depends what mosfet is in use)?

**Comment by masterzen** [comment]
I've pushed a rebased to master version using `writePinHigh`/`writePinLow` instead of the tricky AND/OR versions.

**Comment by drashna** [comment]
Looks like some merge conflicts have occurred. Could you fix that?

**Comment by a-chol** [comment]
After testing again, I confirm that there is an inconsistency between full hardware backlight and software PWM with hardware timer : in HW mode, BL_INC and BL_DEC appropriately increase and decrease backlight. In software mode, BL_INC decreases the intensity while BL_DEC increases it. Can you switch between both modes on your keyboard and confirm that both modes work consistently?

**Comment by masterzen** [comment]
@a-chol I don't have a board supporting hardware PWM. I think this is a problem of `BACKLIGHT_ON_STATE` that is incorrect for your board (it should be `1`) or my code that incorrectly implements the led on/led off.

Would you mind checking `master` (ie without my PR) and see if you get the exact same issue in pure software mode?

If you don't get the issue, then this is a problem in my implementation.

Thanks!

**Comment by a-chol** [comment]
I also get the inconsistent level management between hardware and software mode from master. I tested from the 4015 PR to get proper pin adressing for my kb. From the initial custom backlight model provided for ps2avrgb keyboards, it seems backlight is turned on when the pin is high. 
The semantic of BACKLIGHT_ON_STATE should be discussed as a separate issue. TI would suggest that this is out of the scope of the PR.

**Comment by masterzen** [comment]
@a-chol,

Since I wanted to make sure the code in this PR is correct:
I checked my keyboard, and `BL_INC` increases the backlight level, `BL_DEC` decreases it

Your keyboard is originally a ps2avrgb based keyboard, so I thought I might check this firmware code to look how they turn the led on and off. It [turns out it sets the pin high](https://github.com/showjean/ps2avrU/blob/6aff9e8ed0738d37f282cfdbce8e0346d68d862a/firmware/ps2avrGB/ledconfig.h#L15) to switch the leds on, which in terms of QMK corresponds to [writePinHigh](https://github.com/qmk/qmk_firmware/blob/master/quantum/quantum.h#L152).
Now if you look at [my PR `backlight_on`](https://github.com/qmk/qmk_firmware/blob/58aeeee6e7898abadf3240fffcfdf40cef5f2bbd/quantum/quantum.c#L1101), this means `BACKLIGHT_ON_STATE` should be `1`, same if you look [to the QMK master](https://github.com/qmk/qmk_firmware/blob/1fa6c1d818add2c72381639a836180af5aa86738/quantum/quantum.c#L1107).

To convince myself that any ps2avrgb board should use `BACKLIGHT_ON_STATE=1`, I checked another keyboard. The [jj40 backlight code](https://github.com/qmk/qmk_firmware/blob/master/keyboards/jj40/backlight.c#L141) uses hardware PWM in "non inverted output" mode (see "Table 20-4. Compare Output Mode, Fast PWM Mode" in the Atmega32a datasheet), which means the pin will be set to high from `BOTTOM` to `OCR1`, then low afterwards (which is coherent with the software PWM using `BACKLIGHT_ON_STATE=1`.

So the generated waveform will look like this:
```
| ̅ ̅ ̅|___| ̅ ̅ ̅|___| ̅ ̅ ̅|__
```
instead of:
```
____| ̅ ̅ |____| ̅ ̅ |____| ̅ ̅ |
```
for `BACKLIGHT_ON_STATE=0`.

To summarize, it seems that any ps2avrgb keyboard should use `BACKLIGHT_ON_STATE=1` when using software PWM :)

**Comment by masterzen** [comment]
I really should refresh PR pages before posting a new comment :) I missed your last answer.

Yes I concur, the discussion about the semantics of `BACKLIGHT_ON_STATE` doesn't belong to this PR.

Anyway, thanks for your help and tests !

**Comment by a-chol** [comment]
Is this PR based on PR 4015? If so, it does require that 4015 is merge before this one can be merged, right?
I would suggest the following changes to add support for ps2avrgb boards.

[ps2avrgb_backlight_support.txt](https://github.com/qmk/qmk_firmware/files/2529695/ps2avrgb_backlight_support.txt)

If you don't want to wait for 4015 to be merged, I can still submit a new PR afterwards.

**Comment by masterzen** [comment]
@a-chol,

No this PR doesn't depend on #4105. If #4105 is merged, then there will be support of other atmega than the Atmega32u4, if not only the Atmega32u4 will be supported.

Do you know if only `D4` and `D5` are used on ps2avrgb board for backlight?
Because maybe we could list all pins on all supported MCU that supports hardware backlight at once?

I'll have a look to the different MCU to try to compile such list of available pins for hardware PWM.

**Comment by drashna** [comment]
That would be awesome!   But yes, it should be a separate PR, I think.

**Comment by masterzen** [comment]
@drashna I created the PR #4324 which opens the hardware backlight code to more MCU. For ATmega32A, it can work only with #4015 applied.

**Comment by drashna** [comment]
Well, that PR is applied, so both PRs should be good to go, I think.

**Comment by drashna** [comment]
@jackhumbert @skullydazed @fredizzimo 

This looks ready to go, and just needs to be reviewed.

**Comment by jbarrios** [comment]
I have a xd60 rev3 and I fetched and checkout the PR, the backlights works better than the origin code but it's not perfect (I'll use it without doubt).
I detected the issues that in sometimes without press any key the backlight flicks, more frequently with low backlight level. And with breathing backlight this flicks are more common when starts the cycle

**Comment by masterzen** [comment]
@jbarrios do you have the rgb underglow activated?

I've also noticed that it sometimes flickers when it is activated. I haven't yet found the cause, since it doesn't happen that frequently (it's less than one time per minute). I need to rebuild my firmware with more debug and let it run to "see" why it sometimes jumps...

Those flickers happens because PWM is very sensible to timing, so if we "miss" an interrupt for one cycle the led will stay on and it looks like it flickers.

At low level, it is more problematic because the "turn off" interrupt can happen approximately at the same time as the "turn on" one (I haven't yet understood exactly how it is possible since normally there can be only one interrupt executing at the same time).

**Comment by masterzen** [comment]
@jbarrios , I did more testing today.
Indeed this is coming from the rgblight feature. If you deactivate or use a still color or a very slow rgb underglow effect, then there is no flickering appearing.
If you instead use rgblight with a quickly color changing scheme, then flickering appears.

My explanation is that, all rgblight effects are using `timer_read` and `timer_elapsed`. Those are blocking interrupts to happen while reading the timer. If interrupts are blocked, even for a very short time, my hybrid implementation interrupts that powers on and off the backlight led is either not called or called to late, leaving the leds on for a bit too long. Indeed the frequency of my hybrid interrupt is quite high to be able to support breathing.

Unfortunately, I don't see how this can be fixed, unless rewriting the rgb effects to not depend on `timer_read` but instead depend on a fixed time interrupt (like this backlight code as discussed in this PR).

Meanwhile I suggest using a slow motion RGB effect or no RGB effect at all :)

**Comment by jbarrios** [comment]
Hello, @masterzen, I've been writing without rgblight around the last 3 days and you are right, the flickering has disapear.

Actually i'm using a close case and the underglow have no sense at all, I'll use the backlight with this PR.

Thank you for the explanation and the implementation, maybe someday we will enjoy all full features :)

**Comment by masterzen** [comment]
Thanks @jbarrios for confirming.

On my side I've found the real cause of flicker during rgb underglow changes. Unfortunately to control the rgb lights, the MCU globally disable the interrupts (see https://github.com/qmk/qmk_firmware/blob/master/drivers/avr/ws2812.c#L275) while sending the data to the rgblights. 

So if you have a rapidly changing lighting pattern, the longer the MCU will run with interrupts disabled, and the longer the interrupts are disabled, the higher is the chance to have the backlights leds to stay on after the PWM duty cycle, leading to flickering.

I don't think this is solvable, because we need the backlight interrupts to arrive on time, but you need a precise timing to control the rgb led array. 
Both seems to be exclusive.

**Comment by masterzen** [comment]
Hey @mechmerlin and @drashna, thanks for merging this PR !
I'm going to update #4324 ASAP.

**Comment by al3ph** [comment]
Will this work for the XD87 (pin D0) ? As that has flickering backlight problems, though it's random, and not related to key presses (as far as I can tell), I assume the firmware is servicing USB events and that causes the soft PWM servicing to get briefly delayed causing the flicker.
